rimage/transform/parallel_projection.go

```diff
@@ -95,5 +97,165 @@ func (pp *ParallelProjection) PointCloudToRGBD(cloud pointcloud.PointCloud) (*ri
+// position onto the XZ plane.
+type ParallelProjectionOntoXZWithRobotMarker struct {
+	robotPose *spatialmath.Pose
+}
+
+const (
+	sigmaLevel        = 7    // level of precision for stdev calculation (determined through experimentation)
+	imageHeight       = 1080 // image height
+	imageWidth        = 1080 // image width
+	missThreshold     = 0.44 // percentage at below which any probability point is will be colored a miss
+	hitThreshold      = 0.53 // percentage at above which any probability point is will be colored a hit
+	pointRadius       = 1    // radius of pointcloud point
+	robotMarkerRadius = 5    // radius of robot marker point
+)
+
+// PointCloudToRGBD creates an image of a pointcloud in the XZ plane, scaling the points to a standard image
+// size. It will also add a red marker to the map to represent the location of the robot. The returned depthMap
+// is unused and so will always be nil.
+func (ppRM *ParallelProjectionOntoXZWithRobotMarker) PointCloudToRGBD(cloud pointcloud.PointCloud,
+) (*rimage.Image, *rimage.DepthMap, error) {
+	meta := cloud.MetaData()
+
+	if cloud.Size() == 0 {
+		return nil, nil, errors.New("invalid projection point cloud is empty")
+	}
+
+	var X, Z []float64
+	cloud.Iterate(0, 0, func(pt r3.Vector, data pointcloud.Data) bool {
+		X = append(X, pt.X)
+		Z = append(Z, pt.Z)
+		return true
+	})
+
+	// Calculate max and min range to be represented in the output image and, if needed, cropping it based on
+	// the mean and standard deviation of the X and Z coordinates
+	meanX, err := stats.Mean(X)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to calculate mean of X values on given point cloud")
+	}
+	stdevX, err := stats.StandardDeviation(X)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to calculate stdev of Z values on given point cloud")
+	}
+
+	meanZ, err := stats.Mean(Z)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to calculate mean of Z values on given point cloud")
+	}
+	stdevZ, err := stats.StandardDeviation(Z)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to calculate stdev of Z values on given point cloud")
+	}
+
+	maxX := math.Min(meanX+float64(sigmaLevel)*stdevX, meta.MaxX)
+	minX := math.Max(meanX-float64(sigmaLevel)*stdevX, meta.MinX)
+	maxZ := math.Min(meanZ+float64(sigmaLevel)*stdevZ, meta.MaxZ)
+	minZ := math.Max(meanZ-float64(sigmaLevel)*stdevZ, meta.MinZ)
+
+	// Change the max and min values to ensure the robot marker can be represented in the output image
+	var robotMarker spatialmath.Pose
+	if ppRM.robotPose != nil {
+		robotMarker = *ppRM.robotPose
+		maxX = math.Max(maxX, robotMarker.Point().X)
+		minX = math.Min(minX, robotMarker.Point().X)
+		maxZ = math.Max(maxZ, robotMarker.Point().Z)
+		minZ = math.Min(minZ, robotMarker.Point().Z)
+	}
+
+	// Calculate the scale factors
+	var scaleFactor float64
+	if (maxX-minX) != 0 || (maxZ-minZ) != 0 {
+		widthScaleFactor := float64(imageWidth-1) / (maxX - minX)
```

Comment:
Why are we subtracting 1 from the image width & height here?

Reply:
so that if a point/robot marker defines some of the max/min values it will still be displayed in the image. Pixels in the image range from 0 -> imageWidth-1 and 0 -> imageHeight-1

Reply:
So all points appear in the image. Only valid pixels in the image are 0->imageWidth-1 and 0->imageHeight-1 due to non inclusiveness of upper bound. For large data sets this is not a big deal although even then, if the robot marker's radius = 1 and it helped define one or more of the max bounds, it would not appear on the produced image.